File: Data Structures & Algorithms/trapping-rain-water/submission-12.js
class Solution {
    /**
     * @param {number[]} height
     * @return {number}
     */
    trap(height) { 
        let stack = []  
        let res = 0

        for (let i = 0; i < height.length; i ++) {
            let cur = height[i] 
            while(stack.length > 0 && height[stack[stack.length - 1]] < cur) { 
                const mid = stack.pop() 

                if (stack.length > 0) {   
                    const left = height[stack[stack.length - 1]]  
                    const h = Math.min( cur, left ) 
                    const w = i - stack[stack.length - 1] - 1 

                    res += (h - height[mid]) * w
                }
            }

            stack.push(i)
        } 

        return res
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-12.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        left = [] 
        star = [] 

        for i in range(len(s)): 
            c = s[i] 

            if c == ')': 
                if not left and not star: 
                    return False 

                if left: 
                    left.pop()  
                else: 
                    star.pop() 
                    
            elif c == '(': 
                left.append(i) 
            else: 
                star.append(i) 

        while left and star:
            if left.pop() > star.pop(): 
                return False 

        return not left